data_processors/lims/services.py

```diff
@@ -130,3 +137,142 @@ def __csv_column_to_field_name(column_name: str) -> str:
+
+@transaction.atomic
+def persist_labmetadata(df: pd.DataFrame, rewrite: bool = False) -> Dict[str, int]:
+    """
+    Persist labmetadata from a pandas dataframe into the db
+    :param df: dataframe to persist
+    :param rewrite: whether we are rewriting the data
+    :return: result statistics - count of updated, new and invalid LabMetadata rows
+    """
+    df_starting_len = len(df.index)
+    logger.info(f"Start processing LabMetadata data")
+    if rewrite:
+        # Delete all rows first
+        logger.info("REWRITE MODE: Deleting all existing records")
+        LabMetadata.objects.all().delete()
+
+    # clean the DF and reset the index
+    df = df.applymap(_clean_datacell)
+    df = df.drop_duplicates()
+    df = _remove_rows_with_empty_required_cols(df)
+    df = df.reset_index(drop=True)
+
+    # Below code assumes a labmetadata's unique identifier is a tuple of library_id and sample_name
+
+    # Assemble a df of entries to update by first querying for existing labmetadatas 
+    libids_samplenames = list(zip( df['library_id'].tolist(),  df['sample_name'].tolist() ))
+    query = reduce(operator.or_, (Q(library_id=l, sample_name=s) for l, s in libids_samplenames))
+    existing_labmetadatas = LabMetadata.objects.filter(query)
+    existing_identifier_tuples = tuple((lm.library_id, lm.sample_name) for lm in existing_labmetadatas)
+    df_to_update =  df[pd.Series(list(zip(df['library_id'], df['sample_name']))).isin(existing_identifier_tuples)]
+    
+    # Assemble df of entries to create.
+    df_to_create = df.append(df_to_update)
+    df_to_create = df_to_create[~df_to_create.index.duplicated(keep=False)]
+
+    # Create labmetadata ojbects
+    try:
+        instances_insert = _make_labmetadata_instances_from_dataframe(df_to_create)
+        rows_created = LabMetadata.objects.bulk_create(instances_insert,batch_size=100)
+    except Exception as e:
+        logger.error("Error bulk creating objects! No new rows created")
+        logger.error(e)   
+        logger.debug(traceback.format_exc())
+        rows_created = []
+        raise 
+
+    # Update labmetadatas
+    rows_updated = 0
+    for row in df_to_update.itertuples():
+        rows_updated += _update_labmetadata_instance(row)
+
+    logger.debug("Updated " + str((rows_updated)))
+    logger.debug("New " + str(len(rows_created)))
+    logger.debug("Invalid " + str(len(df.index) - len(rows_created) - rows_updated ))
+
+    return {
+        'labmetadata_row_update_count':rows_updated,
+        'labmetadata_row_new_count': len(rows_created),
+        'labmetadata_row_invalid_count': df_starting_len - len(rows_created) - rows_updated
+    }
+
+def _clean_datacell(value):
+    #python NaNs are != to themselves
+    if value == '-' or value == np.nan or value != value:
+        value = ''
+    if (isinstance(value,str) and value.strip() == ''):
+        value = ''
+    return value
+    
+def _remove_rows_with_empty_required_cols(df):
```

Comment:
Did you come across cases for this?
There should not be records without at least a sample ID *and* a library ID. If there are then it's likely an issue in the data and should probably be flagged, rather than ignored/skipped? Or is that due to legacy data (we may have some odd records there, yes)?
Also, why is sample_name required? This should not be needed. Does that even make it into Clarity?

Reply:
the one case of those variables missing is the 2018 tracking sheet - it has none of those fields. can throw error instead if you think its wiser. 

Re: sample_name: it is the concat of sample_id and external_sample_id (in the tracking sheet). it does get stored in Clarity. i used library_id+sample_name as a "unique idenfitier for a row" throughout the code but its a bit arbitray, in theory it could be library_id+external_sample_id.

im not actually sure what sample_id represents in the sheet - since Clarity ID Generator is meant to generate a field calld sample_id (autoincrement+string except if external_sample_id already has sample_id associated in which case return that).

ought i change the 'unique identifier' to library_id+external_sample_id , or some other combo?

Reply:
A) I'd ignore all 2018 data (that's not conform to our current metadata formatting (in the future we'd either just archive it or we'd need to make it conform to the expected metadata format).

I'd also ignore sample_name for now. One to discuss with the lab if they want to keep it, but it's a compound of already known fields and not used anywhere as far as I know.

The uniqueness story is a whole different can of worms. For now the library ID itself should be unique (according to the rules set out in the GDrive metadata sheet). However that's likely to change soon and will depend a bit on how Clarity models that. Now library ID is the most separating field. It should be unique as is unless there are topups or reruns, in which case the library ID is modified by a `_topup` or `_rerun` extension to keep it unique. We'd like to avoid that in the future (as it's still the exact same library), we just need to figure out how and where to model the topup/rerun information.